### PR TITLE
fix(clusterpools): cluster-pool-ipv4-cidr/cluster-pool-ipv6-cidr

### PR DESCRIPTION
**Pull request by cndoit18**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

```release-note
`cluster-pool-ipv4-cidr` and `cluster-pool-ipv6-cidr` options now accept string slices and not just string
```

### CODE REVIEW

File: operator/flags.go [hunk 1]
```diff
@@ -171,7 +171,7 @@ func init() {
 	flags.Bool(option.EnableIPv4Name, defaults.EnableIPv4, "Enable IPv4 support")
 	option.BindEnv(option.EnableIPv4Name)
 
-	flags.String(operatorOption.ClusterPoolIPv4CIDR, "",
+	flags.StringSlice(operatorOption.ClusterPoolIPv4CIDR, []string{""},
```
Comment (cndoit18) [inline operator/flags.go]: see 	https://github.com/cilium/cilium/blob/f6dcb4ffe4e5933fbdecb9f4d27937ef292c42a6/operator/option/config.go#L292

**Comment by gandro** [comment]
Thanks for the contribution.

I believe however changing this value breaks upgrades. If users use "--reuse-values" when performing "helm upgrade", Helm will likely fail trying to apply `join` to a string, no?

My proposal therefore would be to introduce a new Helm value (e.g. `operator.clusterPoolIPv6PodCIDRList`), modify the chart such that it prefers the new list-based one (if set), but fails back to the old string-based one if not. And then we can deprecate the old one eventually. At least we should mention this change it the update notes.

While less elegant, there is also an option where we don't change the Helm value at all. I believe if you only changed `flags.String` to `flags.StringSlice` that you can specify multiple CIDRs via Helm by just comma-separating them, e.g. "ipam.operator.clusterPoolIPv4PodCIDR=10.0.0.0/8,172.16.0.0/12" should work too.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 2]
```diff
@@ -707,11 +707,11 @@ data:
 
 {{- if eq $ipam "cluster-pool" }}
 {{- if .Values.ipv4.enabled }}
-  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | quote }}
+  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | join " " | quote }}
   cluster-pool-ipv4-mask-size: {{ .Values.ipam.operator.clusterPoolIPv4MaskSize | quote  }}
 {{- end }}
 {{- if .Values.ipv6.enabled }}
-  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | quote }}
+  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | join " " |quote }}
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: I don't think using space works as a separator for pflag.StringSlice.

```suggestion
  cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDR | join "," |quote }}
```

Comment (cndoit18) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: https://github.com/cilium/cilium/blob/f3597588e0bbc29045a669f759cc4aa99dc34953/pkg/option/config.go#L2330

I checked this place and if "," is used, it is judged to be a whole string and not a list of strings.
Because the parameters are read from `ReadDirConfig`

**Comment by cndoit18** [comment]
> clusterPoolIPv6PodCIDRList



> Thanks for the contribution.
> 
> I believe however changing this value breaks upgrades. If users use "--reuse-values" when performing "helm upgrade", Helm will likely fail trying to apply `join` to a string, no?
> 
> My proposal therefore would be to introduce a new Helm value (e.g. `operator.clusterPoolIPv6PodCIDRList`), modify the chart such that it prefers the new list-based one (if set), but fails back to the old string-based one if not. And then we can deprecate the old one eventually. At least we should mention this change it the update notes.
> 
> While less elegant, there is also an option where we don't change the Helm value at all. I believe if you only changed `flags.String` to `flags.StringSlice` that you can specify multiple CIDRs via Helm by just comma-separating them, e.g. "ipam.operator.clusterPoolIPv4PodCIDR=10.0.0.0/8,172.16.0.0/12" should work too.

Thank you very much, I agree with you. I'll think about how to modify it

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 2 above)
Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Ah, you seem to be correct correct! Thanks for checking!

**Comment by cndoit18** [comment]
> Thanks for the contribution.
> 
> I believe however changing this value breaks upgrades. If users use "--reuse-values" when performing "helm upgrade", Helm will likely fail trying to apply `join` to a string, no?
> 
> My proposal therefore would be to introduce a new Helm value (e.g. `operator.clusterPoolIPv6PodCIDRList`), modify the chart such that it prefers the new list-based one (if set), but fails back to the old string-based one if not. And then we can deprecate the old one eventually. At least we should mention this change it the update notes.
> 
> While less elegant, there is also an option where we don't change the Helm value at all. I believe if you only changed `flags.String` to `flags.StringSlice` that you can specify multiple CIDRs via Helm by just comma-separating them, e.g. "ipam.operator.clusterPoolIPv4PodCIDR=10.0.0.0/8,172.16.0.0/12" should work too.

done

File: install/kubernetes/cilium/values.yaml [hunk 3]
```diff
@@ -925,10 +930,14 @@ ipam:
   operator:
     # -- IPv4 CIDR range to delegate to individual nodes for IPAM.
     clusterPoolIPv4PodCIDR: "10.0.0.0/8"
```
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
    # -- Deprecated in favor of ipam.operator.clusterPoolIPv4PodCIDRList.
    # IPv4 CIDR range to delegate to individual nodes for IPAM.
    clusterPoolIPv4PodCIDR: "10.0.0.0/8"
```

File: install/kubernetes/cilium/values.yaml [hunk 4]
```diff
@@ -106,13 +106,13 @@ affinity:
                 - linux
   podAntiAffinity:
     requiredDuringSchedulingIgnoredDuringExecution:
-    - labelSelector:
```
Comment (errordeveloper) [inline install/kubernetes/cilium/values.yaml]: This and other indentation changes to `value.yaml` seem unnecessary. If you would like to cleanup indentation, please put that into a separate commit, or better - another PR :)

File: install/kubernetes/cilium/values.yaml [hunk 5]
```diff
@@ -127,7 +127,8 @@ extraEnv: {}
 extraInitContainers: []
 
 # -- Additional agent hostPath mounts.
-extraHostPathMounts: []
+extraHostPathMounts:
+  []
```
Comment (errordeveloper) [inline install/kubernetes/cilium/values.yaml]: This and similar changes below don't match the style in which this file is currently written...

File: install/kubernetes/cilium/values.yaml (hunk 4 above)
Comment (cndoit18) [reply install/kubernetes/cilium/values.yaml]: Sorry, these are caused by my own formatting tools

File: operator/flags.go [hunk 6]
```diff
@@ -178,7 +178,7 @@ func init() {
 	flags.Bool(option.EnableIPv4Name, defaults.EnableIPv4, "Enable IPv4 support")
 	option.BindEnv(option.EnableIPv4Name)
 
-	flags.String(operatorOption.ClusterPoolIPv4CIDR, "",
+	flags.StringSlice(operatorOption.ClusterPoolIPv4CIDR, []string{},
```
Comment (errordeveloper) [inline operator/flags.go:181]: The change of field type doesn't seem to be included in this commit, did you forget to run `git add`?

**Review by errordeveloper — CHANGES_REQUESTED** [review]
Please see my comments

File: install/kubernetes/cilium/values.yaml [hunk 7]
```diff
@@ -925,10 +930,14 @@ ipam:
   operator:
     # -- IPv4 CIDR range to delegate to individual nodes for IPAM.
     clusterPoolIPv4PodCIDR: "10.0.0.0/8"
+    # -- IPv4 CIDR list range to delegate to individual nodes for IPAM.
+    clusterPoolIPv4PodCIDRList: []
     # -- IPv4 CIDR mask size to delegate to individual nodes for IPAM.
     clusterPoolIPv4MaskSize: 24
     # -- IPv6 CIDR range to delegate to individual nodes for IPAM.
     clusterPoolIPv6PodCIDR: "fd00::/104"
```
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
    # -- Deprecated in favor of ipam.operator.clusterPoolIPv6PodCIDRList.
    # IPv6 CIDR range to delegate to individual nodes for IPAM.
    clusterPoolIPv6PodCIDR: "fd00::/104"
```

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 8]
```diff
@@ -707,11 +707,11 @@ data:
 
 {{- if eq $ipam "cluster-pool" }}
 {{- if .Values.ipv4.enabled }}
-  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | quote }}
+  cluster-pool-ipv4-cidr: {{ append .Values.ipam.operator.clusterPoolIPv4PodCIDRList .Values.ipam.operator.clusterPoolIPv4PodCIDR | join " " | quote }}
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: I think it's better to use an `if-else` statement here:

The use of `append` kind of makes it hard to use non-default values for `.ipam.operator.clusterPoolIPv4PodCIDR`.

For example, if I want to have `[192.168.0.0/16, 172.16.0.0/12]` as my pod CIDRs, naively I would specify the following value: `ipam.operator.clusterPoolIPv4PodCIDRList=[192.168.0.0/16, 172.16.0.0/12]`

But that would be wrong, since this `append` will pull in the default value of  `10.0.0.0/8` too, so I accidentally would end up with `[192.168.0.0/16, 172.16.0.0/12 10.0.0.0/8]`

I would have to set `ipam.operator.clusterPoolIPv4PodCIDR=172.16.0.0/12` **and** `ipam.operator.clusterPoolIPv4PodCIDRList=[192.168.0.0/16]` to achieve my desired result, which I think is a bit confusing.

Therefore, my suggestion would be to use the list if specified, and fall back on the deprecated value if not:
```suggestion
  {{- if .Values.ipam.operator.clusterPoolIPv4PodCIDRList }}
  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDRList | join " " | quote }}
  {{- else }}
  cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDR | quote }}
  {{- end }}
```

**Review by gandro — COMMENTED** [review]
Some inline feedback around the deprecation of the old value.

Thanks a lot!

File: operator/flags.go (hunk 6 above)
Comment (gandro) [reply operator/flags.go:181]: Not necessary. The code actually already reads it as a string slice. See https://github.com/cilium/cilium/pull/17780#discussion_r742735839 and https://github.com/cilium/cilium/blob/858fca7e2cfa78d3b5faa8d28086c2d8581fde8c/operator/option/config.go#L376

Comment (errordeveloper) [reply operator/flags.go:181]: I see, thanks for clarifying.

File: operator/flags.go (hunk 1 above)
Comment (errordeveloper) [reply operator/flags.go]: Ack!

File: install/kubernetes/cilium/values.yaml (hunk 5 above)
Comment (joestringer) [reply install/kubernetes/cilium/values.yaml]: Indeed, was there a particular tool you used to format them this way? I don't think that the previous developers working on this file were aware of any such tool.

Comment (cndoit18) [reply install/kubernetes/cilium/values.yaml]: > Indeed, was there a particular tool you used to format them this way? I don't think that the previous developers working on this file were aware of any such tool.

https://github.com/redhat-developer/yaml-language-server

Yes. my vscode editor will automatically format it

**Review by gandro — APPROVED** [review]
Looks good to me! Awesome, thank you for taking care of this!

**Review by joestringer — CHANGES_REQUESTED** [review]
Thanks for the PR. At a glance this seems like it should be safe for upgrades from existing clusterpool configuration, although I have not tested it out.

There's a few references to the old name for this field in the docs, could you fix them up?

```
$ git grep clusterPoolIPv4
Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4PodCIDR=<IPv4CIDR>``
Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4MaskSize=<IPv4MaskSize>``
Documentation/gettingstarted/k8s-install-openshift-okd.rst:reflect those changes. Namely ``clusterNetwork`` should match ``nativeRoutingCIDR``, ``clusterPoolIPv4PodCIDR`` and ``clusterPoolIPv4MaskSize``.
...
```

Also, should this be added into the `Documentation/operations/upgrade.rst` upgrade guide to inform users that they should migrate their configurations to the new settings?

**Comment by joestringer** [comment]
I suspect that some of the failures in CI are because this PR has not been rebased against the latest master branch recently. Could you also rebasing against master while addressing the above feedback?

**Comment by cndoit18** [comment]
ABC was deprecated in favor of BBC

> Thanks for the PR. At a glance this seems like it should be safe for upgrades from existing clusterpool configuration, although I have not tested it out.
> 
> There's a few references to the old name for this field in the docs, could you fix them up?
> 
> ```
> $ git grep clusterPoolIPv4
> Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4PodCIDR=<IPv4CIDR>``
> Documentation/gettingstarted/ipam-cluster-pool.rst:   * ``--set ipam.operator.clusterPoolIPv4MaskSize=<IPv4MaskSize>``
> Documentation/gettingstarted/k8s-install-openshift-okd.rst:reflect those changes. Namely ``clusterNetwork`` should match ``nativeRoutingCIDR``, ``clusterPoolIPv4PodCIDR`` and ``clusterPoolIPv4MaskSize``.
> ...
> ```
> 
> Also, should this be added into the `Documentation/operations/upgrade.rst` upgrade guide to inform users that they should migrate their configurations to the new settings?

Thx, updated

**Review by joestringer — APPROVED** [review]
LGTM, thanks.

**Comment by cndoit18** [comment]
Hi, Is there anything else I need to do next?

**Comment by gandro** [comment]
> Hi, Is there anything else I need to do next?

Let's give @errordeveloper some time to re-review. Other than that, this should be good to go.

I'll also re-run CI again to in-cooperate the last changes.

**Comment by cndoit18** [comment]
> > Hi, Is there anything else I need to do next?
> 
> Let's give @errordeveloper some time to re-review. Other than that, this should be good to go.
> 
> I'll also re-run CI again to incooperate the last changes.

thank you

**Review by errordeveloper — APPROVED** [review]

**Comment by cndoit18** [comment]
Hi, do I need to do something to solve this CI can not succeed?

**Comment by gandro** [comment]
> Hi, do I need to do something to solve this CI can not succeed?

Pushing new commits unfortunately invalidates previous CI runs and their links, so I can't tell you what went wrong the last time besides one known-flaky test that got flagged (which is unrelated to your changes) here: https://github.com/cilium/cilium/pull/17780#issuecomment-972692672

I'll run the tests again (this can only be done by team members with commit access unfortunately), please avoid pushing to the branch after the tests have run to keep the result validated.

**Comment by cndoit18** [comment]
> > Hi, do I need to do something to solve this CI can not succeed?
> 
> Pushing new commits unfortunately invalidates previous CI runs and their links, so I can't tell you what went wrong the last time besides one known-flaky test that got flagged (which is unrelated to your changes) here: [#17780 (comment)](https://github.com/cilium/cilium/pull/17780#issuecomment-972692672)
> 
> I'll run the tests again (this can only be done by team members with commit access unfortunately), please avoid pushing to the branch after the tests have run to keep the result validated. There is usually no need to rebase.

Thank you very much

**Comment by cndoit18** [comment]
> /test
> 
> Job 'Cilium-PR-K8s-GKE' hit: #17353 (88.79% similarity)
> 
> Job 'Cilium-PR-K8s-1.21-kernel-4.19' hit: #17881 (98.18% similarity)

Sorry, it seems that the rebase is a past commit, I need to fix it.

**Comment by cndoit18** [comment]
> /test

Can you retrigger it? Thank you

**Comment by gandro** [comment]
> > /test
> 
> Can you retrigger it? Thank you

Yes, will re-trigger it as soon as the Image CI Build stage has finished. Please note that it is usually fine if your branch is not up to date with the master branch, there is no need to rebase, unless there is a known bug somewhere.
